keps/sig-node/20190221-maxInitialFailureCount-health-probes.md

```diff
@@ -0,0 +1,152 @@
+
+This proposal adds a numerical option `maxInitialFailureCount` to probes allowing a greater number of failures during the initial start of the container before taking action, while keeping `failureThreshold` at a minimum to restart deadlocked containers after an acceptable delay.
+
+## Motivation
+
+Slow starting containers here refer to containers that require a significant amount of time (one to several minutes) to start. There can be various reasons for this slow startup:
+
+- long data initialization: only the first startup takes a lot of time
+- heavy workload: every startups take a lot of time
+- underpowered/overloaded node: startup times depend on external factors
+
+The main problem with this kind containers is that they should be given enough time to start before having `livenessProbe` fail `failureThreshold` times, which triggers a kill by the `kubelet` before they have a chance to be up.
+
+There are various strategies to handle this situation with the current API:
+
+- Delay the initial `livenessProbe` sufficiently to permit the container to start up (set `initialDelaySeconds` greater than **startup time**). While this ensures no `livenessProbe` will run and fail during the startup period (triggering a kill), it also delays deadlock detection if the container starts faster than `initialDelaySeconds`. Also, since the `livenessProbe` isn't run at all during startup, there is no feedback loop on the actual startup time of the container.
+- Increase the allowed number of `livenessProbe` failures until `kubelet` kills the container (set `failureThreshold` so that `failureThreshold` times `periodSeconds` is greater than **startup time**). While this gives enough time for the container to start up and allows a feedback loop, it prevents the container from being killed in a timely manner if it deadlocks or otherwise hangs after it has initially successfully come up.
+
+However, none of these strategies give an acceptable protection against deadlocks, which is the primary use case of `livenessProbe`.
+
+### Goals
+
+- Allow slow starting containers to run safely during startup with health probes enabled.
+- Improve documentation of the `Probe` structure in core types' API.
+- Improve `kubernetes.io/docs` section about Pod lifecycle:
+  - Clearly state that PostStart handlers do not delay probe executions.
+  - Introduce and explain this new option.
+  - Document that `kubelet` does not save states, and what are the implications with this new option (see Risks and Mitigations).
+  - Document appropriate use cases for this new option.
+
+### Non-Goals
+
+- Make probe parameters (`initialDelaySeconds`, `failureThreshold`, `periodSeconds`, ...) auto-tunable depending on node performance (only provide a way to allow a longer grace period at the start of the container).
```

Comment:
More broadly, this does not address the issue of pod load affecting startup (or any other liveness probe that may be delayed due to load).  This proposal is strictly at the pod level, not the node level.